File: hackathon_CPP_SDK/src-gen/nevonex/gpsplugin/GPSPluginProvider.cpp
/*
 * nevonex/gpsplugin/GPSPluginProvider.cpp
 * Copyright (c) Robert Bosch GmbH. All rights reserved.
 */

#include "GPSPluginProvider.hpp"
#include <ecore/EObject.hpp>
#include <ecore/EClass.hpp>
#include "nevonex/gpsplugin/GpspluginPackage.hpp"
#include <ecorecpp/mapping.hpp>
#include <nevonex-fcal-platform/log/Logger.hpp>

/*PROTECTED REGION ID(GPSPluginProvider.cpp) START*/
// Please, enable the protected region if you add manually written code.
// To do this, add the keyword ENABLED before START.
/*PROTECTED REGION END*/

using namespace ::nevonex::gpsplugin;
using namespace ::nevonex::log;

GPSPluginProvider::GPSPluginProvider() : m_gPSPlugin(0)
{
    _initialize();

    /*PROTECTED REGION ID(GPSPluginProvider__GPSPluginProvider) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

GPSPluginProvider::~GPSPluginProvider()
{
    m_gPSPlugin.reset();
}

// References
::nevonex::gpsplugin::GPSPlugin_ptr GPSPluginProvider::getGPSPlugin() const
{
    return m_gPSPlugin;
}

void GPSPluginProvider::setGPSPlugin(::nevonex::gpsplugin::GPSPlugin_ptr _gPSPlugin)
{
    if (m_gPSPlugin)
        m_gPSPlugin->_setEContainer(GPSPluginProvider_ptr(),
                GpspluginPackage::_instance()->getGPSPluginProvider__gPSPlugin());
    if (_gPSPlugin)
        _gPSPlugin->_setEContainer(_this(),
                GpspluginPackage::_instance()->getGPSPluginProvider__gPSPlugin());
    m_gPSPlugin = _gPSPlugin;
}

// Operations
void GPSPluginProvider::createMachines(std::istream & _stream)
{
    /*PROTECTED REGION ID(GPSPluginProvider__createMachines) START*/
    /*PROTECTED REGION END*/
}

void GPSPluginProvider::constructChildTypes(::ecore::EJavaObject _parent, ::ecore::EObject_ptr _parentObj, ::ecore::EString const& _path)
{
    /*PROTECTED REGION ID(GPSPluginProvider__constructChildTypes) START*/
    /*PROTECTED REGION END*/
}

void GPSPluginProvider::initMachineProvider()
{
    ::nevonex::types::IMachineProvider::initMachineProvider();
    /*PROTECTED REGION ID(GPSPluginProvider__initMachineProvider) START*/
    /*PROTECTED REGION END*/
}

void GPSPluginProvider::stopMachineProvider()
{
    /*PROTECTED REGION ID(GPSPluginProvider__stopMachineProvider) START*/
    /*PROTECTED REGION END*/
}

void GPSPluginProvider::start()
{
    ::nevonex::types::IMachineProvider::start();
    /*PROTECTED REGION ID(GPSPluginProvider__start) START*/
    /*PROTECTED REGION END*/
}

void GPSPluginProvider::stop()
{
    ::nevonex::types::IMachineProvider::stop();
    /*PROTECTED REGION ID(GPSPluginProvider__stop) START*/
    /*PROTECTED REGION END*/
}

void GPSPluginProvider::restart()
{
    ::nevonex::types::IMachineProvider::restart();
    /*PROTECTED REGION ID(GPSPluginProvider__restart) START*/
    /*PROTECTED REGION END*/
}

void GPSPluginProvider::onConnectionStatusChange(const std::string & index, ::nevonex::machine::MachineConnectionInfo_ptr)
{
    /*PROTECTED REGION ID(GPSPluginProvider__onConnectionStatusChange) START*/
    /*PROTECTED REGION END*/
}

// PropertyChange delegation (diamond inheritance requires explicit override)
void GPSPluginProvider::addPropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::addPropertyChangeListener(_listener);
}

void GPSPluginProvider::removePropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::removePropertyChangeListener(_listener);
}

void GPSPluginProvider::notifyPropertyChange(::ecore::EString const &_name,
        ::ecore::EJavaObject const &_oldValue,
        ::ecore::EJavaObject const &_newValue)
{
    using namespace ::nevonex::types;
    PropertyChange::notifyPropertyChange(_name, _oldValue, _newValue);
}

File: com.bosch.fsp.hackathon.gen/src/com/bosch/nevonex/gpsplugin/impl/GPSPluginProvider.java
/**
Copyright (c) Robert Bosch GmbH. All rights reserved.
*/
package com.bosch.nevonex.gpsplugin.impl;

import com.bosch.fsp.logger.FCALLogs;
import com.bosch.fsp.logger.LoggerConstants;

import com.bosch.fsp.runtime.feature.IMachine;
import com.bosch.fsp.runtime.feature.IMachineProvider;
import com.bosch.fsp.runtime.feature.MachineConnectionInfo;

import com.bosch.fsp.runtime.feature.exception.CommunicationException;
import com.bosch.fsp.runtime.feature.exception.MachineInitException;
import com.bosch.fsp.runtime.feature.exception.NevonexException;

import com.bosch.fsp.runtime.util.internal.Util;

import com.bosch.nevonex.common.ITopicObject;

import com.bosch.nevonex.fcb.IFCALController;
import com.bosch.nevonex.fcb.impl.ConnectionFactory;
import com.bosch.nevonex.fcb.impl.FCALController;
import com.bosch.nevonex.fcb.impl.PublishConnectionFactory;

import com.bosch.nevonex.gpsplugin.IGPSPlugin;
import com.bosch.nevonex.gpsplugin.IGPSPluginProvider;

import com.bosch.nevonex.types.IPropertyChange;
import com.bosch.nevonex.types.impl.TypesPackage;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import java.io.InputStream;

import java.lang.reflect.InvocationTargetException;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.commons.lang3.exception.ExceptionUtils;

import org.eclipse.emf.common.notify.NotificationChain;
import org.eclipse.emf.common.util.BasicEList;
import org.eclipse.emf.common.util.EList;
import org.eclipse.emf.common.util.Enumerator;

import org.eclipse.emf.ecore.EClass;
import org.eclipse.emf.ecore.EClassifier;
import org.eclipse.emf.ecore.EObject;
import org.eclipse.emf.ecore.EReference;
import org.eclipse.emf.ecore.EStructuralFeature;
import org.eclipse.emf.ecore.InternalEObject;

import org.eclipse.emf.ecore.impl.EObjectImpl;

import org.eclipse.emf.ecore.util.EDataTypeUniqueEList;

import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

/**
 * An implementation of the model object '<em><b>GPSPluginProvider</b></em>'.
 * @generated
 */
public class GPSPluginProvider extends EObjectImpl implements IGPSPluginProvider, IMachineProvider {
	protected EList<PropertyChangeListener> listeners;
	protected IFCALController controller;
	protected IGPSPlugin gPSPlugin;

	private Map<String, ITopicObject> indexToObjectMap = new HashMap<>();

	/**
	 * @generated
	 */
	public GPSPluginProvider() {
		super();
	}

	public IGPSPlugin getGPSPlugin() {
		if (gPSPlugin != null && ((org.eclipse.emf.ecore.EObject) gPSPlugin).eIsProxy()) {
			org.eclipse.emf.ecore.InternalEObject old = (org.eclipse.emf.ecore.InternalEObject) gPSPlugin;
			gPSPlugin = (IGPSPlugin) eResolveProxy(old);
		}
		return gPSPlugin;
	}

	public IGPSPlugin basicGetGPSPlugin() {
		return gPSPlugin;
	}

	public void setGPSPlugin(IGPSPlugin newGPSPlugin) {
		gPSPlugin = newGPSPlugin;
	}

	public boolean acceptDom(String root) {
		return "gpsplugin".equalsIgnoreCase(root);
	}

	/**
	 * @generated
	 */
	public void createMachines(InputStream stream) throws MachineInitException {
		try {
			Element root = Util.getDomRootElement(stream);
			List<Node> secondLevelNodeList = new ArrayList<Node>();
			if ("root".equalsIgnoreCase(root.getNodeName())) {
				NodeList childNodes = root.getChildNodes();
				for (int i = 0; i < childNodes.getLength(); i++) {
					Node secondLevelNode = childNodes.item(i);
					if (secondLevelNode.getNodeType() == Node.ELEMENT_NODE
							&& acceptDom(secondLevelNode.getNodeName())) {
						secondLevelNodeList.add(secondLevelNode);
					}
				}
			} else {
				if (root.getNodeType() == Node.ELEMENT_NODE && acceptDom(root.getNodeName())) {
					secondLevelNodeList.add(root);
				}
			}
			if (secondLevelNodeList.isEmpty()) {
				return;
			}
			// Machine creation logic
		} catch (Exception e) {
			throw new MachineInitException("Exception while initializing machine data", e);
		}
	}

	/**
	 * @generated
	 */
	public void constructChildTypes(Node parent, Object parentObj, String path) {
		// Child type construction logic
	}

	/**
	 * @generated
	 */
	public void initMachineProvider() throws MachineInitException {
		try {
			this.controller = new FCALController();
			ConnectionFactory.getInstance().getProviders().add(this);
			PublishConnectionFactory.getInstance();
		} catch (Exception e) {
			throw new MachineInitException("Exception while initializing the machine provider.", e);
		}
	}

	/**
	 * @generated
	 */
	public void startMachineProvider() throws MachineInitException {
		try {
			ConnectionFactory.getInstance().initConnectionFactory();
			PublishConnectionFactory.getInstance().initConnectionFactory();
		} catch (CommunicationException e) {
			throw new MachineInitException(e.getMessage(), e);
		}
	}

	/**
	 * @generated
	 */
	public void stopMachineProvider() throws NevonexException {
		try {
			ConnectionFactory.getInstance().terminateConnectionFactory();
			PublishConnectionFactory.getInstance().terminateConnectionFactory();
		} catch (Exception e) {
			throw new NevonexException(e.getMessage(), e);
		}
	}

	/**
	 * @generated
	 */
	public ITopicObject getTopicElement(String index) {
		return indexToObjectMap.get(index);
	}

	/**
	 * @generated
	 */
	public void onConnectionStatusChange(IMachine machine, MachineConnectionInfo info) {
		notifyPropertyChange(machine, MachineConnectionInfo.CONNECT_PROPERTY, null, info);
	}

	public void notifyPropertyChange(Object source, String name, Object oldValue, Object newValue) {
		if (listeners != null) {
			for (PropertyChangeListener listener : this.listeners) {
				listener.propertyChange(new PropertyChangeEvent(source, name, oldValue, newValue));
			}
		}
	}

	/**
	 * @generated
	 */
	public void addPropertyChangeListener(PropertyChangeListener listener) {
		if (listeners == null) {
			listeners = new BasicEList<>();
		}
		listeners.add(listener);
	}

	/**
	 * @generated
	 */
	public void removePropertyChangeListener(PropertyChangeListener listener) {
		if (listeners != null) {
			listeners.remove(listener);
		}
	}

	/**
	 * @generated
	 */
	public void notifyPropertyChange(String name, Object oldValue, Object newValue) {
		if (listeners != null) {
			for (PropertyChangeListener listener : this.listeners) {
				listener.propertyChange(new PropertyChangeEvent(this, name, oldValue, newValue));
			}
		}
	}

	/**
	 * @generated
	 */
	@Override
	protected EClass eStaticClass() {
		return GpspluginPackage.Literals.GPS_PLUGIN_PROVIDER;
	}
} //GPSPluginProvider
